### PR TITLE
feat: add paginated logs

### LINKED ISSUE #954: Pagination / Streaming for Event Logs
**Is your feature request related to a problem? Please describe.**
As event logs can yield more results than an RPC will allows (for example, infura failing when the response has over 10k items), looping over a stream of event logs based on a filter or providing a paginated API would be very useful.

**Describe the solution you'd like**

Gnosis has a potential solution [here](https://github.com/gnosis/ethcontract-rs/blob/39d40e06428df7da6d36318054ca672ff4e2d115/ethcontract/src/log.rs#L321) where `past_logs_pages`/`stream` can return a filter based log stream. 

I think something similar would be the easiest implementation. The draft implementation could start by adding a `get_logs_pages` method on the provider that accepts a `Filter` object. However, I am open to feedback / suggestions in terms of how it's designed.

**Additional context**
Understanding the nuances of how eth_getLogs is handled by the RPC provider may be required and may introduce additional complexity. Below is a list of use cases to research to determine how they differ and how to best handle their use cases: 

- eth on infura
- eth on alchemy
- polygon on infura 
- polygon on alchemy
- bsc on infura

### PR DESCRIPTION
**Pull request by meetmangukiya**
Closes #954 

## Solution

Implement a new stream that breaks requested block range into smaller pages that are loaded one at a time and streamed to the user.

Note: this only solves block limitations, if an rpc provider limits to 1000 blocks for instance, you could set page size to 1000 to solve it. However, if the limitation is on the number of logs(infura), this can't solve it automatically or completely, can try smaller page sizes but because number of logs in arbitrary block ranges is also arbitrary it can be hard to decide on a page size.  

## PR Checklist

- [ ] Added Tests
- [x] Added Documentation
- [x] Updated the changelog

### CODE REVIEW

File: ethers-providers/src/lib.rs [hunk 1]
```diff
@@ -421,6 +424,15 @@ pub trait Middleware: Sync + Send + Debug {
...
     }
 
+    /// Returns a stream of logs are loaded in pages of given page size
+    fn get_logs_paginated<'a>(
+        &'a self,
+        filter: &Filter,
+        page_size: u64,
+    ) -> LogQuery<'a, Self::Provider> {
+        self.inner().get_logs_paginated(filter, page_size)
+    }
```
Comment (gakonst) [inline ethers-providers/src/lib.rs:434]: @prestwich wdyt about unifying this logic in the default get_logs behavior?

**Review by gakonst — COMMENTED** [review]
looks good - i think we need a test to ensure it works appropriately, either with a local node or with a large mainnet/testnet query?

**Comment by meetmangukiya** [comment]
> looks good - i think we need a test to ensure it works appropriately, either with a local node or with a large mainnet/testnet query?

Ah actually i tested with an example seems like i missed committing.

File: ethers-providers/src/log_query.rs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+macro_rules! rewake_with_new_state {
+    ($ctx:ident, $this:ident, $new_state:expr) => {
+        $this.state = $new_state;
+        $ctx.waker().wake_by_ref();
+        return Poll::Pending
+    };
+}
+
+impl<'a, P> Unpin for LogQuery<'a, P> {}
```
Comment (mattsse) [inline ethers-providers/src/log_query.rs:58]: is this sound? adding a type bound to P should work

File: ethers-providers/src/log_query.rs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    pub fn with_page_size(mut self, page_size: u64) -> Self {
+        self.page_size = page_size;
+        self
+    }
+}
+
+macro_rules! rewake_with_new_state {
+    ($ctx:ident, $this:ident, $new_state:expr) => {
+        $this.state = $new_state;
+        $ctx.waker().wake_by_ref();
```
Comment (mattsse) [inline ethers-providers/src/log_query.rs:53]: can we make this work without continuously waking the waker? because this will immediately re-poll the query

**Review by mattsse — CHANGES_REQUESTED** [review]

File: ethers-providers/src/lib.rs (hunk 1 above)
Comment (prestwich) [reply ethers-providers/src/lib.rs:434]: i think it's a pretty reasonable way to smooth over inconsistent RPC paging requirements 🤔

File: ethers-providers/src/log_query.rs (hunk 3 above)
Comment (prestwich) [reply ethers-providers/src/log_query.rs:53]: this macro should basically not be used unless the new state blocks on another future

Comment (meetmangukiya) [reply ethers-providers/src/log_query.rs:53]: yeah only using it in case of polling another future

Comment (meetmangukiya) [reply ethers-providers/src/log_query.rs:53]: I suppose, could also just set the new state and return `Poll::Pending`?

File: ethers-providers/src/log_query.rs (hunk 2 above)
Comment (meetmangukiya) [reply ethers-providers/src/log_query.rs:58]: It should be, right? since we are storing a ref to provider, not a raw pointer?

Comment (meetmangukiya) [reply ethers-providers/src/log_query.rs:58]: Removing this impl, seems unnecessary

File: ethers-providers/src/log_query.rs (hunk 3 above)
Comment (mattsse) [reply ethers-providers/src/log_query.rs:53]: > yeah only using it in case of polling another future

got it, that works!

**Review by gakonst — APPROVED** [review]
LGTM. Separately, I wonder if we should unify this with `get_logs`. Would be a breaking change, requiring to change the function signature to return `LogQuery` instead of `Vec<Log>` but that could be navigable as it'd basically only introduce an additional `await` when iterating over the vector of logs.

Is there any benefit to using `get_log` over always using paginated? Probably not?

**Comment by meetmangukiya** [comment]
Yeah, no real benefits of using `get_log`, didn't change it to not break things

**Comment by gakonst** [comment]
Would give the follow-up a try, I think this is a good enough change that it's worth breaking, and we can include in the changelog

File: ethers-providers/src/log_query.rs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+                let filter = self.filter.clone().from_block(from_block).to_block(to_block);
+                let provider = self.provider;
+                // load first page of logs
+                let fut = Box::pin(async move { provider.get_logs(&filter).await });
+                rewake_with_new_state!(ctx, self, LogQueryState::LoadLogs(fut));
+            }
+            LogQueryState::LoadLogs(fut) => {
+                let logs = futures_util::ready!(fut.as_mut().poll(ctx))
+                    .expect("error occurred loading logs");
```
Comment (recmo) [inline ethers-providers/src/log_query.rs:97]: Panicking on failed RPCs is not very nice. I'd much prefer if this was a TryStream instead. Will submit PR when I get to it.

File: ethers-providers/src/log_query.rs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                        // load new logs if there are still more pages to go through
+                        let from_block = self.from_block.unwrap();
+                        let to_block = from_block + self.page_size;
+
+                        // no more pages to load, and everything is consumed
+                        if from_block > self.last_block.unwrap() {
+                            return Poll::Ready(None)
+                        }
+                        // load next page
+                        self.from_block = Some(to_block);
```
Comment (philsippl) [inline ethers-providers/src/log_query.rs:117]: This should be `to_block + 1`, otherwise you double-sync blocks on the boundaries.

Comment (gakonst) [reply ethers-providers/src/log_query.rs:117]: Cc @meetmangukiya

Comment (meetmangukiya) [reply ethers-providers/src/log_query.rs:117]: will verify this

Comment (meetmangukiya) [reply ethers-providers/src/log_query.rs:117]: @philsippl is correct, will fix this

Comment (meetmangukiya) [reply ethers-providers/src/log_query.rs:117]: https://github.com/gakonst/ethers-rs/pull/1328/commits/86c349701b13dd2a248395a0b38e6094752d089f

File: ethers-providers/src/log_query.rs (hunk 4 above)
Comment (meetmangukiya) [reply ethers-providers/src/log_query.rs:97]: Implemented `TryStream` along with other fixes here https://github.com/gakonst/ethers-rs/pull/1328/commits/0143cafc9f187feab49559e107e894c5725854e6

**Pull request merged**